fs: add rm method (#35494)
nodejs/node: doc/api/fs.md

`recursive` option is deprecated, `ENOTDIR` and `ENOENT` will be thrown in
the future.

## `fs.rm(path[, options], callback)`
<!-- YAML

> added: v15.0.0

Reviewer: ```suggestion
added: REPLACEME
```

-->

* `path` {string|Buffer|URL}
* `options` {Object}

>   * `force` Ignore errors

Reviewer 2: nit: I think we might want to be more  specific about the types of errors it ignores, from the `man` page:

> Ignore nonexistent files and missing operands

☝️ perhaps use that as a basis.

>   * `force` Ignore errors

Reviewer 2: ```suggestion
  * `force` Ignore nonexistent path. 
```

  * `maxRetries` {integer} If an `EBUSY`, `EMFILE`, `ENFILE`, `ENOTEMPTY`, or
    `EPERM` error is encountered, Node.js will retry the operation with a linear
    backoff wait of `retryDelay` ms longer on each try. This option represents
    the number of retries. This option is ignored if the `recursive` option is
    not `true`. **Default:** `0`.
  * `recursive` {boolean} If `true`, perform a recursive removal. In
    recursive mode operations are retried on failure. **Default:** `false`.
  * `retryDelay` {integer} The amount of time in milliseconds to wait between
    retries. This option is ignored if the `recursive` option is not `true`.
    **Default:** `100`.
* `callback` {Function}
  * `err` {Error}

> Asynchronous rm(2). No arguments other than a possible exception are given

Reviewer 3: I _think_ this should be `rm(1)`.

It seems that most docs referencing man page definitions refer to system calls. As this one isn't, maybe it ought to be more formally/fully described?

Reviewer 2: > As this one isn't, maybe it ought to be more formally/fully described?

@bengl @iansu I agree, since the goal hear is to provide an API surface analogous to `rm`, but we're not actually using the underlying `rm` system call, perhaps we could provide a description, perhaps motivated by `rm`, but not referencing the system call directly?

Reviewer 4: The `fs` docs start with this:

> The `fs` module enables interacting with the file system in a way modeled on standard POSIX functions.

Perhaps we can...uh...model the text here on the "modeled on" phrasing above? We can say this method's behavior is "modeled on the POSIX `rm` command" or something like that?

> Asynchronous rm(2). No arguments other than a possible exception are given
> to the completion callback.

Reviewer 2: ```suggestion
Asynchronously removes files and directories (modeled on the standard POSIX `rm` utility). No arguments other than a possible exception are given to the completion callback.
```

## `fs.rmSync(path[, options])`
<!-- YAML

> added: v15.0.0

Reviewer: ```suggestion
added: REPLACEME
```

-->

* `path` {string|Buffer|URL}
* `options` {Object}
  * `force` Ignore errors
  * `maxRetries` {integer} If an `EBUSY`, `EMFILE`, `ENFILE`, `ENOTEMPTY`, or
    `EPERM` error is encountered, Node.js will retry the operation with a linear
    backoff wait of `retryDelay` ms longer on each try. This option represents
    the number of retries. This option is ignored if the `recursive` option is
    not `true`. **Default:** `0`.
  * `recursive` {boolean} If `true`, perform a recursive directory removal. In
    recursive mode operations are retried on failure. **Default:** `false`.
  * `retryDelay` {integer} The amount of time in milliseconds to wait between
    retries. This option is ignored if the `recursive` option is not `true`.
    **Default:** `100`.

> Synchronous rm(2). Returns `undefined`.

Reviewer 2: ```suggestion
Synchronously removes files and directories (modeled on the standard POSIX `rm` utility). Returns `undefined`.
```

`recursive` option is deprecated, `ENOTDIR` and `ENOENT` will be thrown in
the future.

## `fs.rm(path[, options], callback)`
<!-- YAML
added: REPLACEME
-->

* `path` {string|Buffer|URL}
* `options` {Object}
  * `force` don't error on nonexistent path
  * `maxRetries` {integer} If an `EBUSY`, `EMFILE`, `ENFILE`, `ENOTEMPTY`, or
    `EPERM` error is encountered, Node.js will retry the operation with a linear

>     backoff wait of `retryDelay` ms longer on each try. This option represents

Reviewer 5: from the sidelines: any particular reason this uses the abbreviated `ms` rather than typing out `milliseconds`? I see the `rm*` methods in these docs uses `ms`, but they are the only ones - all others use the full `milliseconds`

Author: No specific reason. This particular text was just copied from recursive rmdir. I can change it here or do it in my follow up PR to add the deprecation warning to recursive rmdir.

    retries. This option is ignored if the `recursive` option is not `true`.
    **Default:** `100`.
* `callback` {Function}
  * `err` {Error}

Asynchronously removes files and directories (modeled on the standard POSIX `rm`
utility). No arguments other than a possible exception are given to the
completion callback.

## `fs.rmSync(path[, options])`
<!-- YAML
added: REPLACEME
-->

* `path` {string|Buffer|URL}
* `options` {Object}
  * `force` Ignore errors
  * `maxRetries` {integer} If an `EBUSY`, `EMFILE`, `ENFILE`, `ENOTEMPTY`, or
    `EPERM` error is encountered, Node.js will retry the operation with a linear
    backoff wait of `retryDelay` ms longer on each try. This option represents
    the number of retries. This option is ignored if the `recursive` option is
    not `true`. **Default:** `0`.
  * `recursive` {boolean} If `true`, perform a recursive directory removal. In
    recursive mode operations are retried on failure. **Default:** `false`.
  * `retryDelay` {integer} The amount of time in milliseconds to wait between
    retries. This option is ignored if the `recursive` option is not `true`.
    **Default:** `100`.

> Syncronously removes files and directories (modeled on the standard POSIX `rm`

Reviewer 6: ```suggestion
Synchronously removes files and directories (modeled on the standard POSIX `rm`
```

Author: Good catch. I even have a spellcheck in VSCode but it obviously failed me here.
